Index: build/steditlib_vc9.vcproj
===================================================================
--- build/steditlib_vc9.vcproj	(revision 1078)
+++ build/steditlib_vc9.vcproj	(working copy)
@@ -353,6 +353,10 @@
 				>
 			</File>
 			<File
+				RelativePath="..\include\wx\stedit\stedoc.h"
+				>
+			</File>
+			<File
 				RelativePath="..\include\wx\stedit\steexprt.h"
 				>
 			</File>
@@ -453,22 +457,18 @@
 				>
 			</File>
 			<File
-				RelativePath="..\src\stedoc.cpp"
+				RelativePath="..\src\stedlgs.cpp"
 				>
 			</File>
 			<File
-				RelativePath="..\src\wxext.cpp"
+				RelativePath="..\src\stedlgs_wdr.cpp"
 				>
 			</File>
 			<File
-				RelativePath="..\src\stedlgs.cpp"
+				RelativePath="..\src\stedoc.cpp"
 				>
 			</File>
 			<File
-				RelativePath="..\src\stedlgs_wdr.cpp"
-				>
-			</File>
-			<File
 				RelativePath="..\src\steexprt.cpp"
 				>
 			</File>
@@ -516,6 +516,10 @@
 				RelativePath="..\src\stestyls.cpp"
 				>
 			</File>
+			<File
+				RelativePath="..\src\wxext.cpp"
+				>
+			</File>
 		</Filter>
 	</Files>
 	<Globals>
Index: include/wx/stedit/stedefs.h
===================================================================
--- include/wx/stedit/stedefs.h	(revision 1078)
+++ include/wx/stedit/stedefs.h	(working copy)
@@ -29,6 +29,8 @@
     #include "wx/stedit/setup.h"
 #endif // !defined(_STESETUP_H_)
 
+#define STEDIT_DOCVIEW 1
+
 //-----------------------------------------------------------------------------
 // The version of wxStEdit
 //-----------------------------------------------------------------------------
Index: include/wx/stedit/stedit.h
===================================================================
--- include/wx/stedit/stedit.h	(revision 1078)
+++ include/wx/stedit/stedit.h	(working copy)
@@ -21,7 +21,6 @@
 class WXDLLIMPEXP_FWD_CORE wxFindDialogEvent;
 class WXDLLIMPEXP_FWD_CORE wxToolBar;
 class WXDLLIMPEXP_FWD_BASE wxConfigBase;
-class WXDLLIMPEXP_FWD_CORE wxFileHistory;
 class WXDLLIMPEXP_FWD_CORE wxDocument;
 
 #include "wx/stedit/stedefs.h"
@@ -94,11 +93,12 @@
 //  You normally do not need to access any of this, use the member functions
 //  in the wxSTEditor.
 //-----------------------------------------------------------------------------
+class WXDLLIMPEXP_FWD_STEDIT wxSTEditorDoc;
 class WXDLLIMPEXP_STEDIT wxSTEditorRefData : public wxObjectRefData
 {
 public:
     wxSTEditorRefData();
-    virtual ~wxSTEditorRefData() { m_editors.Clear(); }
+    virtual ~wxSTEditorRefData();
 
     // Find/Add/Remove editors that share this data
     size_t GetEditorCount() const { return m_editors.GetCount(); }
@@ -114,7 +114,6 @@
     // implementation
     wxArrayPtrVoid m_editors;       // editors that share this data
 
-    wxString     m_fileName;        // current filename for the editor
     wxDateTime   m_modifiedTime;    // file modification time, else invalid
     wxTreeItemId m_treeItemId;      // the treeitem if tracked in a wxTreeCtrl
 
@@ -131,6 +130,14 @@
     wxSTEditorPrefs  m_stePrefs;
     wxSTEditorStyles m_steStyles;
     wxSTEditorLangs  m_steLangs;
+
+#if STEDIT_DOCVIEW
+    wxSTEditorDoc* m_doc;
+#else
+    wxString     m_fileName;        // current filename for the editor
+#endif
+    wxString GetFilename() const;
+    void SetFilename(const wxString&);
 };
 
 //-----------------------------------------------------------------------------
@@ -679,6 +686,9 @@
     // Get the ref counted data for the editor (ref counted for splitting)
     //  The ref data is ALWAYS expected to exist, do NOT call wxObject::UnRef
     wxSTEditorRefData* GetSTERefData() const { return (wxSTEditorRefData*)GetRefData(); }
+#if STEDIT_DOCVIEW
+    wxDocument* GetDocument() const { return (wxDocument*)GetSTERefData()->m_doc; }
+#endif
 
     // If this editor is going to use a Refed document, run this after construction
     //  to have this mirror the original wxSTEditor, origEditor isn't modified
Index: include/wx/stedit/stedoc.h
===================================================================
--- include/wx/stedit/stedoc.h	(revision 0)
+++ include/wx/stedit/stedoc.h	(revision 0)
@@ -0,0 +1,60 @@
+///////////////////////////////////////////////////////////////////////////////
+// Name:        stedoc.h
+// Purpose:     docview interaction
+// Author:      Troels K
+// Modified by:
+// Created:     2009-11-11
+// Copyright:   (c) John Labenski, Troels K
+// Licence:     wxWidgets licence
+///////////////////////////////////////////////////////////////////////////////
+
+#if STEDIT_DOCVIEW
+
+class WXDLLIMPEXP_FWD_STEDIT wxSTEditorFrame;
+class WXDLLIMPEXP_FWD_STEDIT wxSTEditorRefData;
+
+//-----------------------------------------------------------------------------
+// wxSTEditorDocManager
+//-----------------------------------------------------------------------------
+
+class WXDLLIMPEXP_STEDIT wxSTEditorDocManager : public wxDocManager
+{
+    typedef wxDocManager base;
+protected:
+    wxSTEditorFrame* m_frame;
+public:
+    wxSTEditorDocManager(wxSTEditorFrame*);
+
+    virtual wxDocument* CreateDocument(const wxString& path, long flags = 0);
+};
+
+//-----------------------------------------------------------------------------
+// wxSTEditorDoc
+//-----------------------------------------------------------------------------
+
+class WXDLLIMPEXP_STEDIT wxSTEditorDoc : public wxDocument
+{
+    typedef wxDocument base;
+protected:
+    wxSTEditorRefData* m_editor;
+public:
+    wxSTEditorDoc(wxSTEditorRefData*);
+
+    static wxDocTemplate* GetDocTemplate(wxDocManager* manager = NULL);
+
+    wxWindow* GetModalParent() { return GetDocumentWindow(); }
+
+    bool InitDocument();
+    bool OnSaveModified();
+
+public:
+    virtual ~wxSTEditorDoc();
+    virtual bool Save();
+    virtual bool SaveAs();
+
+    friend class wxSTEditorView;
+};
+
+/////////////////////////////////////////////////////////////////////////////
+
+#endif // STEDIT_DOCVIEW
Index: include/wx/stedit/steframe.h
===================================================================
--- include/wx/stedit/steframe.h	(revision 1078)
+++ include/wx/stedit/steframe.h	(working copy)
@@ -23,9 +23,9 @@
 class WXDLLIMPEXP_FWD_CORE wxKeyEvent;
 class WXDLLIMPEXP_FWD_CORE wxToolBar;
 class WXDLLIMPEXP_FWD_BASE wxConfigBase;
-class WXDLLIMPEXP_FWD_CORE wxFileHistory;
 class WXDLLIMPEXP_FWD_CORE wxTreeCtrl;
 class WXDLLIMPEXP_FWD_CORE wxTreeEvent;
+class WXDLLIMPEXP_FWD_STEDIT wxSTEditorDocManager;
 
 #include "wx/stedit/stedefs.h"
 
@@ -61,16 +61,24 @@
 //                   ^
 //              SideSplitter
 
-class WXDLLIMPEXP_STEDIT wxSTEditorFrame : public wxFrame
+class WXDLLIMPEXP_STEDIT wxSTEditorFrame : 
+#if STEDIT_DOCVIEW
+public wxDocParentFrame
 {
+   typedef wxDocParentFrame base;
+#else
+public wxFrame
+{
+   typedef wxFrame base;
+#endif
 public:
-    wxSTEditorFrame() : wxFrame() { Init(); }
-    wxSTEditorFrame(wxWindow *parent, wxWindowID id,
+    wxSTEditorFrame() : base() { Init(); }
+    wxSTEditorFrame(wxFrame* parent, wxWindowID id,
                     const wxString& title = wxT(STE_APPDISPLAYNAME),
                     const wxPoint& pos = wxDefaultPosition,
                     const wxSize& size = wxSize(400, 400),
                     long style = wxDEFAULT_FRAME_STYLE,
-                    const wxString& name = wxT("wxSTEditorFrame")) : wxFrame()
+                    const wxString& name = wxT("wxSTEditorFrame")) : base()
     {
         Init();
         Create(parent, id, title, pos, size, style, name);
@@ -79,7 +87,7 @@
     virtual ~wxSTEditorFrame();
     virtual bool Destroy();
 
-    bool Create(wxWindow *parent, wxWindowID id,
+    bool Create(wxFrame *parent, wxWindowID id,
                 const wxString& title = wxT(STE_APPDISPLAYNAME),
                 const wxPoint& pos = wxDefaultPosition,
                 const wxSize& size = wxDefaultSize,
Index: samples/stedit/wxstedit.cpp
===================================================================
--- samples/stedit/wxstedit.cpp	(revision 1078)
+++ samples/stedit/wxstedit.cpp	(working copy)
@@ -38,6 +38,7 @@
 #ifndef WX_PRECOMP
     #include "wx/wx.h"
 #endif
+#include "wx/docview.h"
 
 #if (wxVERSION_NUMBER >= 2900)
 #include "wx/stc/private.h" // stc2wx()
@@ -288,6 +289,7 @@
             return false;
         }
     }
+    //steOptions.GetMenuManager()->SetMenuOptionType(STE_MENU_NOTEBOOK, false);
 
     // Remove the Help menu since wxMac will pull out the wxID_ABOUT to add to
     // the system menu and then hide the Help menu. Later on when we add items
Index: src/stedit.cpp
===================================================================
--- src/stedit.cpp	(revision 1078)
+++ src/stedit.cpp	(working copy)
@@ -47,6 +47,7 @@
 
 #include "wx/stedit/stedit.h"
 #include "wx/stedit/steexprt.h"
+#include "wx/stedit/stedoc.h"
 
 //-----------------------------------------------------------------------------
 // Global data
@@ -79,8 +80,37 @@
                                       m_last_autoindent_len(0),
                                       m_steLang_id(STE_LANG_NULL)
 {
+#if STEDIT_DOCVIEW
+   m_doc = new wxSTEditorDoc(this);
+#endif
 }
 
+wxSTEditorRefData::~wxSTEditorRefData()
+{
+    m_editors.Clear();
+#if STEDIT_DOCVIEW
+    //wxDELETE(m_doc);
+#endif
+}
+
+wxString wxSTEditorRefData::GetFilename() const
+{
+#if STEDIT_DOCVIEW
+   return m_doc->GetFilename();
+#else
+   return m_fileName;
+#endif
+}
+
+void wxSTEditorRefData::SetFilename(const wxString& filename)
+{
+#if STEDIT_DOCVIEW
+   m_doc->SetFilename(filename);
+#else
+   m_fileName = filename;
+#endif
+}
+
 //-----------------------------------------------------------------------------
 // wxSTEditor
 //-----------------------------------------------------------------------------
@@ -728,6 +758,13 @@
     m_activating = true;
     IsAlteredOnDisk(true);
     m_activating = false;
+
+#if STEDIT_DOCVIEW
+    if (GetDocument()->GetFirstView())
+    {
+        GetDocument()->GetFirstView()->Activate(true);
+    }
+#endif
 }
 
 bool wxSTEditor::PositionToXY(long pos, long *col, long *row)
@@ -2060,14 +2097,14 @@
 
 wxString wxSTEditor::GetFileName() const
 {
-    return GetSTERefData()->m_fileName;
+    return GetSTERefData()->GetFilename();
 }
 
 void wxSTEditor::SetFileName(const wxString &fileName, bool send_event)
 {
-    if (GetSTERefData()->m_fileName != fileName)
+    if (GetSTERefData()->GetFilename() != fileName)
     {
-        GetSTERefData()->m_fileName = fileName;
+        GetSTERefData()->SetFilename(fileName);
         if (send_event)
             SendEvent(wxEVT_STE_STATE_CHANGED, STE_FILENAME, GetState(), fileName);
     }
Index: src/stedoc.cpp
===================================================================
--- src/stedoc.cpp	(revision 0)
+++ src/stedoc.cpp	(revision 0)
@@ -0,0 +1,139 @@
+///////////////////////////////////////////////////////////////////////////////
+// Name:        stedoc.cpp
+// Purpose:     docview interaction
+// Author:      Troels K
+// Modified by:
+// Created:     2009-11-11
+// Copyright:   (c) John Labenski, Troels K
+// Licence:     wxWidgets licence
+///////////////////////////////////////////////////////////////////////////////
+
+#include "precomp.h"
+
+#include "wx/stedit/stedefs.h"
+
+#if STEDIT_DOCVIEW
+#include "wx/stedit/stedoc.h"
+#include "wx/stedit/stedit.h"
+#include "wx/stedit/steopts.h"
+#include "wx/stedit/steframe.h"
+#include "wx/stedit/steprint.h"
+
+//-----------------------------------------------------------------------------
+// wxSTEditorDocManager
+//-----------------------------------------------------------------------------
+
+wxSTEditorDocManager::wxSTEditorDocManager(wxSTEditorFrame* frame) : wxDocManager(), m_frame(frame)
+{
+}
+
+wxDocument* wxSTEditorDocManager::CreateDocument(const wxString& path, long flags)
+{
+     //return base::CreateDocument(path, flags);
+     wxDocument* doc = NULL;
+     wxSTEditorNotebook* notebook = m_frame->GetEditorNotebook();
+     wxSTEditor* editor = m_frame->GetEditor();
+
+     bool ok;
+     if (flags & wxDOC_NEW)
+     {
+         ok = notebook ? notebook->NewPage(path) : editor->NewFile(path);
+     }
+     else
+     {
+         ok = notebook ? notebook->LoadFile(path) : editor->LoadFile(path);
+     }
+     if (ok)
+     {
+         doc = m_frame->GetEditor()->GetDocument();
+         ok = wxStaticCast(doc, wxSTEditorDoc)->InitDocument();
+         if (!ok)
+         {
+            wxDELETE(doc);
+         }
+     }
+     return doc;
+}
+
+//-----------------------------------------------------------------------------
+// wxSTEditorDoc
+//-----------------------------------------------------------------------------
+
+wxSTEditorDoc::wxSTEditorDoc(wxSTEditorRefData* editor) : wxDocument(), m_editor(editor)
+{
+}
+
+wxSTEditorDoc::~wxSTEditorDoc()
+{
+   m_editor->m_doc = NULL;
+}
+
+bool wxSTEditorDoc::InitDocument()
+{
+   bool ok = GetDocTemplate()->InitDocument(this, GetFilename());
+   if (ok)
+   {
+       GetFirstView()->Activate(true);
+   }
+   return ok;
+}
+
+bool wxSTEditorDoc::Save()
+{
+   return m_editor->GetEditor(0)->SaveFile(false);
+}
+
+bool wxSTEditorDoc::SaveAs()
+{
+   return m_editor->GetEditor(0)->SaveFile(true);
+}
+
+//-----------------------------------------------------------------------------
+// wxSTEditorView
+//-----------------------------------------------------------------------------
+
+class wxSTEditorView : public wxView
+{
+    typedef wxView base;
+    DECLARE_DYNAMIC_CLASS(wxSTEditorView)
+public:
+    wxSTEditorView() : wxView()
+    {
+    }
+    virtual ~wxSTEditorView()
+    {
+    }
+    virtual void OnDraw(wxDC*)
+    {
+    }
+    virtual wxPrintout* OnCreatePrintout();
+};
+
+IMPLEMENT_DYNAMIC_CLASS(wxSTEditorView, wxView)
+
+wxPrintout* wxSTEditorView::OnCreatePrintout()
+{
+    wxSTEditorDoc* doc = wxStaticCast(GetDocument(), wxSTEditorDoc);
+    return new wxSTEditorPrintout(doc->m_editor->GetEditor(0));
+}
+
+bool wxSTEditorDoc::OnSaveModified()
+{
+    return true; // avoid additional prompt
+}
+
+/*static*/ wxDocTemplate* wxSTEditorDoc::GetDocTemplate(wxDocManager* docManager)
+{
+    static wxDocTemplate* doctemplate = NULL;
+    if (docManager)
+    {
+       doctemplate = new wxDocTemplate(docManager, wxT("Text Files"), wxT("*.")wxT("txt"), 
+           wxT(""), wxT("txt"), wxT("text doc"), wxT("text view"),
+           CLASSINFO(wxSTEditorDoc), CLASSINFO(wxSTEditorView));
+    }
+    return doctemplate;
+}
+
+/////////////////////////////////////////////////////////////////////////////
+
+#endif // STEDIT_DOCVIEW
Index: src/steframe.cpp
===================================================================
--- src/steframe.cpp	(revision 1078)
+++ src/steframe.cpp	(working copy)
@@ -13,6 +13,7 @@
 
 #include "wx/stedit/stedit.h"
 #include "wx/stedit/steframe.h"
+#include "wx/stedit/stedoc.h"
 
 #include "../art/pencil16.xpm"
 #include "../art/pencil32.xpm"
@@ -22,9 +23,9 @@
 //-----------------------------------------------------------------------------
 // wxSTEditorFrame
 //-----------------------------------------------------------------------------
-IMPLEMENT_DYNAMIC_CLASS(wxSTEditorFrame, wxFrame)
+IMPLEMENT_DYNAMIC_CLASS(wxSTEditorFrame, base)
 
-BEGIN_EVENT_TABLE(wxSTEditorFrame, wxFrame)
+BEGIN_EVENT_TABLE(wxSTEditorFrame, base)
     EVT_MENU_OPEN             (wxSTEditorFrame::OnMenuOpen)
     EVT_MENU                  (wxID_ANY, wxSTEditorFrame::OnMenu)
 
@@ -49,9 +50,14 @@
     m_fileTreeCtrl     = NULL;
     m_sideSplitterWin1 = NULL;
     m_sideSplitterWin2 = NULL;
+
+#if STEDIT_DOCVIEW
+    m_docManager = new wxSTEditorDocManager(this);
+    wxSTEditorDoc::GetDocTemplate(m_docManager);
+#endif
 }
 
-bool wxSTEditorFrame::Create(wxWindow *parent, wxWindowID id,
+bool wxSTEditorFrame::Create(wxFrame *parent, wxWindowID id,
                              const wxString& title,
                              const wxPoint& pos, const wxSize& size,
                              long  style,
@@ -59,7 +65,11 @@
 {
     m_titleBase = title;
 
-    if (!wxFrame::Create(parent, id, title, pos, size, style, name))
+#if STEDIT_DOCVIEW
+    if (!base::Create(m_docManager, parent, id, title, pos, size, style, name))
+#else
+    if (!base::Create(parent, id, title, pos, size, style, name))
+#endif
         return false;
 
     // Set the frame's icons
@@ -94,7 +104,11 @@
         GetOptions().GetFindReplaceData())
         GetOptions().GetFindReplaceData()->SaveConfig(*config,
                       GetOptions().GetConfigPath(STE_OPTION_CFGPATH_FINDREPLACE));
+#if STEDIT_DOCVIEW
+    delete GetDocumentManager();
+#endif
 }
+
 bool wxSTEditorFrame::Destroy()
 {
     SetSendSTEEvents(false);
@@ -105,7 +119,7 @@
     if (GetStatusBar() && (GetStatusBar() == GetOptions().GetStatusBar()))
         GetOptions().SetStatusBar(NULL);
 
-    return wxFrame::Destroy();
+    return base::Destroy();
 }
 void wxSTEditorFrame::SetSendSTEEvents(bool send)
 {
@@ -311,7 +325,12 @@
     // if we've got an editor let it update gui
     wxSTEditor *editor = GetEditor();
     if (editor)
+    {
+    #if STEDIT_DOCVIEW
+        wxStaticCast(editor->GetDocument(), wxSTEditorDoc)->InitDocument();
+    #endif
         editor->UpdateAllItems();
+    }
 }
 
 wxSTEditor *wxSTEditorFrame::GetEditor(int page) const
@@ -604,6 +623,13 @@
     event.Skip();
     wxSTEditor *editor = event.GetEditor();
 
+#if STEDIT_DOCVIEW
+    if ( event.HasStateChange(STE_MODIFIED) )
+    {
+       editor->GetDocument()->Modify(editor->IsModified());
+    }
+#endif
+
     if ( event.HasStateChange(STE_FILENAME | STE_MODIFIED) )
     {
         wxString modified = editor->IsModified() ? wxT("*") : wxT("");
@@ -719,11 +745,14 @@
         if (GetOptions().GetFileHistory())
         {
             wxString fileName = GetOptions().GetFileHistory()->GetHistoryFile(win_id-wxID_FILE1);
-
+         #if STEDIT_DOCVIEW
+            m_docManager->CreateDocument(fileName);
+         #else
             if (GetEditorNotebook())
                 GetEditorNotebook()->LoadFile(fileName);
             else if (editor)
                 editor->LoadFile(fileName);
+         #endif
         }
 
         return true;
